File: belezza-api/src/test/scala/com/belezza/api/performance/BasicSimulation.scala
package com.belezza.api.performance

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

/**
 * Basic load test simulation for Belezza API.
 *
 * This simulation tests:
 * - Authentication endpoints
 * - Metrics endpoints
 * - Concurrent user scenarios
 *
 * Run with: mvn gatling:test
 */
class BasicSimulation extends Simulation {

  // Configuration
  val baseUrl = System.getProperty("baseUrl", "http://localhost:8080")
  val users = Integer.getInteger("users", 100)
  val duration = Integer.getInteger("duration", 60).seconds

  // HTTP Protocol Configuration
  val httpProtocol = http
    .baseUrl(baseUrl)
    .acceptHeader("application/json")
    .contentTypeHeader("application/json")
    .userAgentHeader("Gatling/LoadTest")

  // Scenarios

  /**
   * Scenario 1: Login flow
   * Simulates users logging in
   */
  val loginScenario = scenario("Login Flow")
    .exec(http("Login")
      .post("/api/auth/login")
      .body(StringBody("""{"email":"test@example.com","password":"password"}"""))
      .check(status.is(200))
      .check(jsonPath("$.accessToken").saveAs("token"))
    )
    .pause(1)

  /**
   * Scenario 2: Get user profile
   * Requires authentication
   */
  val profileScenario = scenario("Get Profile")
    .exec(http("Login")
      .post("/api/auth/login")
      .body(StringBody("""{"email":"test@example.com","password":"password"}"""))
      .check(status.is(200))
      .check(jsonPath("$.accessToken").saveAs("token"))
    )
    .pause(1)
    .exec(http("Get Profile")
      .get("/api/auth/me")
      .header("Authorization", "Bearer ${token}")
      .check(status.is(200))
    )

  /**
   * Scenario 3: Get metrics
   * Simulates salon owners checking their metrics
   */
  val metricsScenario = scenario("Get Metrics")
    .exec(http("Login")
      .post("/api/auth/login")
      .body(StringBody("""{"email":"admin@example.com","password":"password"}"""))
      .check(status.is(200))
      .check(jsonPath("$.accessToken").saveAs("token"))
    )
    .pause(1)
    .exec(http("Get Scheduling Metrics")
      .get("/api/metricas/agendamentos")
      .queryParam("salonId", "1")
      .header("Authorization", "Bearer ${token}")
      .check(status.is(200))
    )
    .pause(2)
    .exec(http("Get Financial Metrics")
      .get("/api/metricas/faturamento")
      .queryParam("salonId", "1")
      .header("Authorization", "Bearer ${token}")
      .check(status.is(200))
    )
    .pause(2)
    .exec(http("Get Social Metrics")
      .get("/api/metricas/social")
      .queryParam("salonId", "1")
      .header("Authorization", "Bearer ${token}")
      .check(status.is(200))
    )

  /**
   * Scenario 4: Health check
   * Simulates monitoring systems
   */
  val healthCheckScenario = scenario("Health Check")
    .exec(http("Health Check")
      .get("/actuator/health")
      .check(status.is(200))
    )
    .pause(5)

  // Load injection strategies

  /**
   * Setup 1: Gradual ramp-up
   * Slowly increases load to test system under increasing pressure
   */
  val gradualRampUp = setUp(
    loginScenario.inject(
      rampUsers(users / 4) during (duration / 4)
    ),
    profileScenario.inject(
      rampUsers(users / 4) during (duration / 4)
    ),
    metricsScenario.inject(
      rampUsers(users / 4) during (duration / 4)
    ),
    healthCheckScenario.inject(
      constantUsersPerSec(2) during duration
    )
  ).protocols(httpProtocol)
    .assertions(
      global.responseTime.max.lt(5000),  // Max response time < 5s
      global.successfulRequests.percent.gt(95)  // 95% success rate
    )

  /**
   * Setup 2: Spike test
   * Sudden burst of users to test system resilience
   */
  val spikeTest = setUp(
    metricsScenario.inject(
      nothingFor(5.seconds),
      atOnceUsers(users)  // All users at once
    )
  ).protocols(httpProtocol)
    .assertions(
      global.responseTime.max.lt(10000),  // Allow higher response time for spike
      global.failedRequests.percent.lt(10)  // Allow some failures during spike
    )

  /**
   * Setup 3: Sustained load
   * Constant load over time to test stability
   */
  val sustainedLoad = setUp(
    loginScenario.inject(
      constantUsersPerSec(2) during duration
    ),
    metricsScenario.inject(
      constantUsersPerSec(5) during duration
    )
  ).protocols(httpProtocol)
    .assertions(
      global.responseTime.mean.lt(2000),  // Mean response time < 2s
      global.successfulRequests.percent.gt(99)  // 99% success rate
    )

  // Default: Use gradual ramp-up
  gradualRampUp
}

File: belezza-api/src/test/java/com/belezza/api/performance/BasicSimulation.java
package com.belezza.api.performance;

import io.gatling.javaapi.core.*;
import io.gatling.javaapi.http.*;

import java.time.Duration;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

/**
 * Gatling performance test simulation for Belezza API.
 *
 * Scenarios:
 * - T9.4.2: 100 concurrent users
 * - T9.4.3: 1000 requests/minute
 * - T9.4.4: Peak load simulation (Monday morning appointments)
 *
 * Run with: mvn gatling:test
 */
public class BasicSimulation extends Simulation {

    // HTTP protocol configuration
    HttpProtocolBuilder httpProtocol = http
            .baseUrl("http://localhost:8080")
            .acceptHeader("application/json")
            .contentTypeHeader("application/json")
            .userAgentHeader("Gatling/Belezza-Performance-Test");

    // Feeders for test data
    FeederBuilder<String> userFeeder = csv("users.csv").random();

    // ====================================
    // Scenario 1: Health Check (Basic)
    // ====================================
    ScenarioBuilder healthCheckScenario = scenario("Health Check Scenario")
            .exec(
                http("Health Check")
                    .get("/actuator/health")
                    .check(status().is(200))
            );

    // ====================================
    // Scenario 2: Authentication Flow
    // ====================================
    ScenarioBuilder authFlowScenario = scenario("Authentication Flow")
            .exec(
                // Login
                http("Login")
                    .post("/api/auth/login")
                    .body(StringBody("""
                        {
                            "email": "test@example.com",
                            "password": "Test123!"
                        }
                        """))
                    .check(status().is(200))
                    .check(jsonPath("$.accessToken").saveAs("token"))
            )
            .pause(Duration.ofMillis(500))
            .exec(
                // Access protected resource
                http("Get Current User")
                    .get("/api/auth/me")
                    .header("Authorization", "Bearer #{token}")
                    .check(status().is(200))
            );

    // ====================================
    // Scenario 3: Metrics Dashboard
    // ====================================
    ScenarioBuilder metricsScenario = scenario("Metrics Dashboard")
            .exec(
                // Login first
                http("Login for Metrics")
                    .post("/api/auth/login")
                    .body(StringBody("""
                        {
                            "email": "admin@example.com",
                            "password": "Admin123!"
                        }
                        """))
                    .check(status().in(200, 401)) // 401 if user doesn't exist
                    .check(jsonPath("$.accessToken").optional().saveAs("token"))
            )
            .doIf(session -> session.contains("token"))
            .then(
                exec(
                    http("Get Scheduling Metrics")
                        .get("/api/metricas/agendamentos?salonId=1")
                        .header("Authorization", "Bearer #{token}")
                        .check(status().in(200, 404))
                )
                .pause(Duration.ofMillis(200))
                .exec(
                    http("Get Financial Metrics")
                        .get("/api/metricas/faturamento?salonId=1")
                        .header("Authorization", "Bearer #{token}")
                        .check(status().in(200, 404))
                )
                .pause(Duration.ofMillis(200))
                .exec(
                    http("Get Social Metrics")
                        .get("/api/metricas/social?salonId=1")
                        .header("Authorization", "Bearer #{token}")
                        .check(status().in(200, 404))
                )
            );

    // ====================================
    // Scenario 4: Peak Monday Morning Load
    // Simulates appointment creation and lookup
    // ====================================
    ScenarioBuilder peakLoadScenario = scenario("Peak Monday Morning")
            .exec(
                // Login
                http("Login")
                    .post("/api/auth/login")
                    .body(StringBody("""
                        {
                            "email": "client@example.com",
                            "password": "Client123!"
                        }
                        """))
                    .check(status().in(200, 401))
                    .check(jsonPath("$.accessToken").optional().saveAs("token"))
            )
            .doIf(session -> session.contains("token"))
            .then(
                // Browse services
                exec(
                    http("List Services")
                        .get("/api/servicos?salonId=1")
                        .header("Authorization", "Bearer #{token}")
                        .check(status().in(200, 404))
                )
                .pause(Duration.ofSeconds(1), Duration.ofSeconds(2))
                // Check availability
                .exec(
                    http("Check Availability")
                        .get("/api/profissionais/1/disponibilidade")
                        .header("Authorization", "Bearer #{token}")
                        .check(status().in(200, 404))
                )
                .pause(Duration.ofMillis(500))
                // List existing appointments
                .exec(
                    http("List My Appointments")
                        .get("/api/agendamentos?clienteId=1")
                        .header("Authorization", "Bearer #{token}")
                        .check(status().in(200, 404))
                )
            );

    // ====================================
    // Load Configurations
    // ====================================

    /**
     * T9.4.2 - 100 concurrent users scenario
     */
    PopulationBuilder scenario100Users = healthCheckScenario.injectOpen(
            rampUsers(100).during(Duration.ofSeconds(30))
    );

    /**
     * T9.4.3 - 1000 requests/minute (approximately 17 req/sec)
     */
    PopulationBuilder scenario1000ReqPerMin = healthCheckScenario.injectOpen(
            constantUsersPerSec(17).during(Duration.ofMinutes(1))
    );

    /**
     * T9.4.4 - Peak load: Monday morning simulation
     * Ramps up from 0 to 200 users over 5 minutes, maintains for 5 minutes, then ramps down
     */
    PopulationBuilder scenarioPeakLoad = peakLoadScenario.injectOpen(
            rampUsers(50).during(Duration.ofMinutes(1)),
            constantUsersPerSec(10).during(Duration.ofMinutes(2)),
            rampUsersPerSec(10).to(50).during(Duration.ofMinutes(2)),
            constantUsersPerSec(50).during(Duration.ofMinutes(3)),
            rampUsersPerSec(50).to(5).during(Duration.ofMinutes(2))
    );

    // ====================================
    // Simulation Setup
    // ====================================

    {
        setUp(
            // Run all scenarios
            healthCheckScenario.injectOpen(
                atOnceUsers(10),
                rampUsers(50).during(Duration.ofSeconds(30))
            ),
            authFlowScenario.injectOpen(
                rampUsers(20).during(Duration.ofSeconds(30))
            ),
            metricsScenario.injectOpen(
                rampUsers(10).during(Duration.ofSeconds(30))
            )
        )
        .protocols(httpProtocol)
        .assertions(
            // Global assertions
            global().responseTime().mean().lt(2000),      // Mean response time < 2s
            global().responseTime().percentile3().lt(3000), // 95th percentile < 3s
            global().successfulRequests().percent().gt(95.0) // >95% success rate
        );
    }
}
